File: demo/src/main/java/com/example/demo/services/ApplicationService.java
package com.example.demo.services;

import com.example.demo.models.Application;
import com.example.demo.models.Candidate;
import com.example.demo.models.Job;
import com.example.demo.models.NotificationType;
import com.example.demo.repositories.ApplicationRepository;
import com.example.demo.repositories.CandidateRepository;
import com.example.demo.repositories.JobRepository;
import com.example.demo.repositories.UserRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;

@Service
public class ApplicationService {

    private final ApplicationRepository applicationRepository;
    private final CandidateRepository candidateRepository;
    private final JobRepository jobRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;
    private final SocketService socketService;

    public ApplicationService(ApplicationRepository applicationRepository, CandidateRepository candidateRepository, JobRepository jobRepository, UserRepository userRepository, NotificationService notificationService, SocketService socketService) {
        this.applicationRepository = applicationRepository;
        this.candidateRepository = candidateRepository;
        this.jobRepository = jobRepository;
        this.userRepository = userRepository;
        this.notificationService = notificationService;
        this.socketService = socketService;
    }

    public Application applyJob(String userEmail, String jobId, String cvSnapshotUrl) {
        Candidate candidate = candidateRepository.findByUser(userRepository.findByEmail(userEmail).orElseThrow()).orElseThrow(() -> new RuntimeException("Candidate profile not found"));
        Job job = jobRepository.findById(jobId).orElseThrow(() -> new RuntimeException("Job not found"));

        if (applicationRepository.findByJobAndCandidate(job, candidate).isPresent()) {
            throw new RuntimeException("Already applied to this job");
        }

        if (job.getCompany().getOwner().getEmail().equals(userEmail)) {
            throw new RuntimeException("Cannot apply to your own job");
        }

        Application application = Application.builder()
                .job(job)
                .candidate(candidate)
                .cvSnapshotUrl(cvSnapshotUrl)
                .build();
        applicationRepository.save(application);

        job.setApplicationsNum(job.getApplicationsNum() + 1);
        jobRepository.save(job);

        // Create notification for the company
        Map<String, Object> notificationData = new HashMap<>();
        notificationData.put("to", job.getCompany().getOwner().getId());
        notificationData.put("from", candidate.getId());
        notificationData.put("fromModel", "Candidate");
        notificationData.put("type", NotificationType.NEW_APPLICATION);
        notificationData.put("displayName", candidate.getFullName());
        notificationData.put("jobTitle", job.getTitle());

        notificationService.createNotification(
            job.getCompany().getOwner(),
            candidate.getId(),
            "Candidate",
            NotificationType.NEW_APPLICATION,
            candidate.getFullName(),
            job.getTitle(),
            null
        );
        socketService.emitNotification(job.getCompany().getOwner().getId(), notificationData);

        return application;
    }

    public boolean cancelApply(String userEmail, String jobId) {
        Candidate candidate = candidateRepository.findByUser(userRepository.findByEmail(userEmail).orElseThrow()).orElseThrow(() -> new RuntimeException("Candidate profile not found"));
        Job job = jobRepository.findById(jobId).orElseThrow(() -> new RuntimeException("Job not found"));

        Application application = applicationRepository.findByJobAndCandidate(job, candidate).orElseThrow(() -> new RuntimeException("Application not found"));
        applicationRepository.delete(application);

        job.setApplicationsNum(job.getApplicationsNum() - 1);
        jobRepository.save(job);
        return true;
    }

    public Map<String, Object> getMyApplications(String userEmail, int page, int limit) {
        Candidate candidate = candidateRepository.findByUser(userRepository.findByEmail(userEmail).orElseThrow()).orElseThrow(() -> new RuntimeException("Candidate profile not found"));

        Pageable pageable = PageRequest.of(page - 1, limit, Sort.by(Sort.Direction.DESC, "createdAt"));
        Page<Application> applicationsPage = applicationRepository.findByCandidate(candidate, pageable);

        Map<String, Object> response = new HashMap<>();
        response.put("data", applicationsPage.getContent());
        Map<String, Object> pagination = new HashMap<>();
        pagination.put("total", applicationsPage.getTotalElements());
        pagination.put("page", applicationsPage.getNumber() + 1);
        pagination.put("limit", applicationsPage.getSize());
        pagination.put("totalPages", applicationsPage.getTotalPages());
        response.put("pagination", pagination);
        return response;
    }
}

File: backend/src/services/application.service.js
import Application from "../models/Application.model.js";
import Candidate from "../models/Candidate.model.js";
import Job from "../models/Job.model.js";
import Company from "../models/Company.model.js";
import { createNotification } from "./notification.service.js";
import { getIo } from "./socket.service.js"; // Import getIo for socket emission

export const applyJob = async (userId, jobId, cvSnapshotUrl) => {
  const candidate = await Candidate.findOne({ userId });
  if (!candidate) throw new Error("Candidate profile not found");

  const existing = await Application.findOne({
    jobId,
    candidateId: candidate._id
  });
  if (existing) {
    throw new Error("Already applied to this job");
  }
  const job = await Job.findById(jobId).populate('companyId');
  if (!job) {
    throw new Error("Job not found");
  }
  const company = job.companyId;
  if(company.ownerId.toString() === userId) {
    throw new Error("Cannot apply to your own job");
  }
  const application = await Application.create({
    jobId,
    candidateId: candidate._id,
    cvSnapshotUrl
  });

  await Job.findByIdAndUpdate(jobId, {
    $inc: { applicationsNum: 1 }
  });

  // Create notification for the company
  const newNotification = await createNotification({
    to: company.ownerId,
    from: candidate._id,
    fromModel: 'Candidate',
    type: 'NEW_APPLICATION',
    displayName: candidate.fullName, // New field: candidate's full name
    jobTitle: job.title, // New field: job title
  });
  const io = getIo();
  if(io){
    io.to(company.ownerId.toString()).emit('new_notification', newNotification);
  }
  return application;
};

export const cancelApply = async (userId, jobId) => {
  const candidate = await Candidate.findOne({ userId });
  if (!candidate) throw new Error("Candidate profile not found");

  const deleted = await Application.findOneAndDelete({
    jobId,
    candidateId: candidate._id
  });

  if (!deleted) {
    throw new Error("Application not found");
  }

  await Job.findByIdAndUpdate(jobId, {
    $inc: { applicationsNum: -1 }
  });

  return true;
};

export const getMyApplications = async (userId, query) => {
  const candidate = await Candidate.findOne({ userId });
  if (!candidate) throw new Error("Candidate profile not found");

  const page = Number(query.page) || 1;
  const limit = Number(query.limit) || 10;
  const skip = (page - 1) * limit;

  const [applications, total] = await Promise.all([
    Application.find({ candidateId: candidate._id })
      .populate({
        path: "jobId",
        select: "title level companyId",
        populate: {
          path: "companyId",
          select: "name logo"
        }
      })
      .sort({ createdAt: -1 })
      .skip(skip)
      .limit(limit),
    Application.countDocuments({ candidateId: candidate._id })
  ]);

  return {
    data: applications,
    pagination: {
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit)
    }
  };
};

export const getApplicationsByJobId = async (jobId) => {
    const applications = await Application.find({ jobId }).populate({
        path: "candidateId",
        select: "fullname cv",
    });
    return applications;
}
